### PR TITLE
Add support for out and inout parameters.

### PR DESCRIPTION
**Pull request by NiiRoZz**
Fix https://github.com/NazaraEngine/ShaderLang/issues/45

### CODE REVIEW

File: include/NZSL/Ast/Nodes.hpp [hunk 1]
```diff
@@ -117,10 +117,18 @@ namespace nzsl::Ast
...
+			In,
+			Out,
+			InOut,
+		};
+
 		NodeType GetType() const override;
 		void Visit(ExpressionVisitor& visitor) override;
 
 		std::vector<ExpressionPtr> parameters;
+		std::vector<ParameterSemantic> parametersSemantic;
```
Comment (SirLynix) [inline include/NZSL/Ast/Nodes.hpp:131]: I think they should be part of the same vector, as each parameter will have its own semantic

File: include/NZSL/Ast/Nodes.hpp [hunk 2]
```diff
@@ -117,10 +117,18 @@ namespace nzsl::Ast
 
 	struct NZSL_API CallFunctionExpression : Expression
 	{
+		enum class ParameterSemantic : Nz::UInt8
```
Comment (SirLynix) [inline include/NZSL/Ast/Nodes.hpp:120]: I'm not sure this enum should be part of CallFunctionExpression

File: src/NZSL/Ast/AstSerializer.cpp [hunk 3]
```diff
@@ -83,6 +83,10 @@ namespace nzsl::Ast
 		Container(node.parameters);
 		for (auto& param : node.parameters)
 			Node(param);
+
+		Container(node.parametersSemantic);
```
Comment (SirLynix) [inline src/NZSL/Ast/AstSerializer.cpp:87]: For retrocompatibility we should check the binary version of the deserializer (IsVersionGreaterThan)

File: tests/src/Tests/FunctionsTests.cpp [hunk 4]
```diff
@@ -210,6 +210,139 @@ OpFunction
...
+	let mainColor: vec3[f32] = vec3[f32](1.0, 1.0, 1.0);
+	let inValue: f32 = 2.0;
+	let inValue2: f32 = 1.0;
+	Half(inout mainColor, out output.value2, inValue, inValue2);
+	output.value = mainColor.x;
+	return output;
+}
+)");
+
+		ExpectSPIRV(*shaderModule, R"(
```
Comment (SirLynix) [inline tests/src/Tests/FunctionsTests.cpp:297]: I'm not sure SPIRV handling is as simple as that, you should use full SPIR-V (with parameter) generation to test it

**Review by SirLynix — CHANGES_REQUESTED** [review]
Thanks a lot.

File: include/NZSL/Ast/Nodes.hpp (hunk 2 above)
Comment (NiiRoZz) [reply include/NZSL/Ast/Nodes.hpp]: Moved out the enum

File: src/NZSL/Ast/AstSerializer.cpp (hunk 3 above)
Comment (NiiRoZz) [reply src/NZSL/Ast/AstSerializer.cpp]: Fixed

**Comment by SirLynix** [comment]
Thanks a lot!

**Pull request merged**